File: bin/far.py
"""
A simple utility that reformats paragraphs to a fixed width.
1. Minimize the variance of the lengths of each line...
2. ...subject to the constraint that the number of lines is optimal
3. Ignore the last line, while making sure it's shorter than average
That's it! Runs in O(NK) where N = # characters and K = width
"""
import sys

PREFIX = set(" >:-*|#$%'\"") # characters allowed to be in a prefix

def get_lines(par: list, width: int) -> list:
    """ Compute optimal line lengths with forward greedy. """
    lines, i, count = [0], 0, 0
    while i < len(par):
        x, v = 0, len(par[i])
        count += 1
        while i + 1 < len(par) and v <= width:
            i += 1
            x, v = v, v + 1 + len(par[i])
            lines.append(count)
        if v <= width:
            i += 1
            lines.append(count)

    return lines

def vardp(par: list, lines: list, width: int) -> list:
    """ Computes the minimum variance, constrained to use optimal lines. """
    # state (index, variance, sum of x^2 terms, sum of x)
    dp = [None]*(len(par) + 1)
    dp[0] = (0, 0, 0, 0)
    for i in range(1, len(par) + 1):
        k, best, sum_x2, sum_x, x = 0, float("inf"), 0, 0, 0
        for j in range(i - 1, -1, -1):
            # add 1 for space, if the current line isn't empty
            v = x + (x != 0) + len(par[j])
            if v <= width:
                x = v
                _, _, sum_x2j, sum_xj = dp[j]
                n = 1 + lines[j]
                sum_x2j += x*x
                sum_xj += x
                # Var[X] = E[X^2] - E[X]^2
                mean = sum_xj/n
                var = sum_x2j/n - mean*mean
                if var < best and n == lines[i]:
                    k, best, sum_x2, sum_x = j, var, sum_x2j, sum_xj
            else:
                break
        dp[i] = (k, best, sum_x2, sum_x)

    return dp

def process(par: list, width: int, prefix: str) -> str:
    """ Takes in a paragraph and returns a string with a new line width. """
    if len(par) == 0:
        return ""
    assert max(map(len, par)) <= width, "line too long"

    lines = get_lines(par, width)
    dp = vardp(par, lines, width)
    # if the paragraph is less than 3 lines long, don't ignore the last line
    if lines[-1] <= 3:
        k = len(par)
    else:
        best, k, x = [float("inf")]*2, [0]*2, 0
        for i in range(len(par) - 1, -1, -1):
            x += (x != 0) + len(par[i])
            if x > width:
                break
            b = x <= dp[i][-1]/lines[i]
            if lines[i] + 1 == lines[-1] and dp[i][1] < best[b]:
                best[b], k[b] = dp[i][1], i
        k = k[1] if best[1] != float("inf") else k[0]

    out, i = [], k
    if k < len(par):
        out.append(" ".join(par[k:]))
    while i > 0:
        j = dp[i][0]
        out.append(" ".join(par[j:i]))
        i = j
    # add prefix to each line
    return "\n".join(map(lambda x: prefix + x, out[::-1]))

def parse_prefix(lines: list) -> tuple:
    """ Parses lines into a list of tokens, taking into account prefixes. """
    # find prefix, where a prefix is defined as a series
    # of the same character, if the character is in PREFIX
    prefix = []
    smallest = min(map(len, lines))
    for ch in range(smallest):
        for line in lines:
            if line[ch] != lines[0][ch] or line[ch] not in PREFIX:
                break
        else:
            prefix.append(lines[0][ch])
            continue
        break
    prefix = "".join(prefix)

    par = []
    for line in lines:
        par += line[len(prefix):].split()

    return par, prefix

if __name__ == "__main__":
    # read command line arguments - one parameter, width
    width = int(sys.argv[1]) if len(sys.argv) > 1 else 79

    # read input into paragraph blocks, making empty lines []
    pars, lines = [], []
    for line in sys.stdin:
        if line != "\n":
            lines.append(line)
        else:
            if len(lines) > 0:
                par, prefix = parse_prefix(lines)
                pars.append((par, width - len(prefix), prefix))
            pars.append(([], width, ""))
            par, lines = [], []
    if len(lines) > 0:
        par, prefix = parse_prefix(lines)
        pars.append((par, width - len(prefix), prefix))

    print("\n".join(map(lambda x: process(*x), pars)))

File: bin/far.cpp
/* clang++ -O3 far.cpp -o far
 * A simple utility that reformats paragraphs to a fixed width.
 * 1. Minimize the variance of the lengths of each line...
 * 2. ... subject to the constraint that the number of lines is optimal
 * 3. Ignore the last line, while making sure it's shorter than average
 * That's it! Runs in O(NK) where N = # characters and K = width
 */
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <unordered_set>
#include <cassert>
#include <limits>
using namespace std;

// renames
#define append push_back
#define last back()
#define mp make_pair
#define fi first
#define se second
#define el '\n'
#define arg const &
#define list vector
// change to long if you run out of memory
typedef long long ll;
typedef string str;
typedef list<ll> li;
typedef list<string> ls;
typedef pair<ll, ll> pi;
// misc
#define For(x, n) for (auto &x : n)
#define print(x) cout << (x) << '\n'
#define len(l) ((int) l.size())
#define in(x, s) (s.find(x) != s.end())

// globals
const ll INF = numeric_limits<ll>::max(); // infinity, but not really
// characters allowed to be in a prefix
const unordered_set<char> PREFIX = {' ', '>', ':', '-', '*',
  '|', '#', '$', '%', '"', '\''};

// fraction methods
inline ll det(pi arg x, pi arg y) { return x.fi*y.se - x.se*y.fi; }    // ad-bc
inline pi mul(pi arg x, pi arg y) { return mp(x.fi*y.fi, x.se*y.se); } // x*y
inline pi sub(pi arg x, pi arg y) { return mp(det(x, y), x.se*y.se); } // x - y
inline bool cmp(pi arg x, pi arg y) { return det(x, y) < 0; }          // x < y

pair<li, ll> get_lines(ls arg para, ll WIDTH) {
  /* Compute optimal line lengths with forward greedy. */
  ll i, num, x, v, chars; i = 0; num = 0; chars = 0;
  li lines = {0};
  while (i < len(para)) {
    x = 0; v = len(para[i]); num++;
    while (v <= WIDTH and i + 1 < len(para)) {
      i++; x = v; v += 1 + len(para[i]);
      lines.append(num);
    }
    chars += x;
    if (v <= WIDTH) {
      i++; chars += v - x;
      lines.append(num);
    }
  }
  return mp(lines, chars);
}

list<pi> vardp(ls arg para, li arg lines, ll WIDTH) {
  /* Computes the minimum variance, constrained to use optimal lines.
   * Minimizing variance is equivalent to minimizing the sum of squares,
   * if the number of lines is constrained. */
  list<pi> dp = {mp(0, 0)}; ll i, j, x, v, k, kj, sum_x2, sum_x2j;
  for (i = 1; i < len(para) + 1; i++) {
    k = 0; x = 0; sum_x2 = INF;
    for (j = i - 1; j >= 0; j--) {
      v = x + (x != 0) + len(para[j]);
      if (v <= WIDTH) {
        x = v;
        kj = dp[j].fi; sum_x2j = dp[j].se;
        sum_x2j += x*x;
        if (sum_x2j < sum_x2 and lines[j] + 1 == lines[i]) {
          k = j; sum_x2 = sum_x2j;
        }
      }
      else {
        break;
      }
    }
    dp.append(mp(k, sum_x2));
  }
  return dp;
}

ll process_dp(ls arg para, li arg lines, ll chars, list<pi> arg dp, ll WIDTH) {
  /* Finds an ending index to minimize variance, ignoring the last line. */
  // if the paragraph is less than three lines long, don't ignore the last line
  if (lines.last <= 3) {
    return len(para);
  }
  ll i, x, kl, kg; pi bestl, bestg, mean, var;
  // find best last line starting point, initialize best to infinity (1/0)
  // maintain two solutions: last line shorter than average, last line greater
  bestl = mp(1, 0); bestg = mp(1, 0); kl = 0; kg = 0; x = 0;
  for (i = len(para) - 1; i >= 0; i--) {
    x += (x != 0) + len(para[i]);
    if (x > WIDTH) {
      break;
    }
    // Var[X] = E[X^2] - E[X]^2
    mean = mp(chars - x, lines[i]);
    var = sub(mp(dp[i].se, lines[i]), mul(mean, mean));
    if (lines[i] + 1 == lines.last) {
      if (det(mp(x, 1), mean) <= 0) {
        if (cmp(var, bestl)) {
          bestl = var; kl = i;
        }
      }
      else {
        if (cmp(var, bestg)) {
          bestg = var; kg = i;
        }
      }
    }
  }
  // use shorter last line if it exists, otherwise default to greater
  return (bestl.se != 0) ? kl : kg;
}

pair<ls, str> parse_prefix(ls arg lines, ll WIDTH) {
  /* Parses lines into a list of tokens, taking into account prefixes. */
  // find prefix, where a prefix is defined as a series
  // of the same character, if the character is in PREFIX
  ls para; str prefix; ll ch; bool end = false;
  for (ch = 0; ch < len(lines[0]); ch++) {
    if (not in (lines[0][ch], PREFIX)) {
      break;
    }
    For(line, lines) {
      if (ch >= len(line) or line[ch] != lines[0][ch]) {
        end = true;
        break;
      }
    }
    if (end) {
      break;
    }
    else {
      prefix += lines[0][ch];
    }
  }
  // remove prefix from each line and load into tokens
  WIDTH -= len(prefix);
  For(line, lines) {
    istringstream ss(line.substr(len(prefix)));
    if (len(line) != 0) {
      str token;
      while (ss >> token) {
        if (len(token) > WIDTH) {
          print("AssertionError: word too long: " + token);
        }
        assert(len(token) <= WIDTH);
        para.append(token);
      }
    }
  }
  return mp(para, prefix);
}

void process(ls arg lines, ll WIDTH) {
  /* Processes lines into a final paragraph and prints it out. */
  ls para; str prefix; list<pi> dp; ll i, j, k, chars; li line_lengths, out;
  tie(para, prefix) = parse_prefix(lines, WIDTH);    // process prefix
  WIDTH -= len(prefix);                              // don't include prefix
  tie(line_lengths, chars) = get_lines(para, WIDTH); // generate line lengths
  dp = vardp(para, line_lengths, WIDTH);             // run dynamic programming
  k = process_dp(para, line_lengths, chars, dp, WIDTH); // process output

  // generate indexes in reverse order
  out = {k};
  while (out.last > 0) {
    out.append(dp[out.last].fi);
  }
  // print out each line, given by contiguous segments of the paragraph
  for (i = len(out) - 1; i > 0; i--) {
    cout << prefix;
    for (j = out[i]; j < out[i - 1] - 1; j++) {
      cout << para[j] << ' ';
    }
    print(para[out[i - 1] - 1]);
  }
  // print last line, if it exists
  if (k < len(para)) {
    cout << prefix;
    for (i = k; i < len(para) - 1; i++) {
      cout << para[i] << ' ';
    }
    print(para.last);
  }
}

int main(int argc, char* argv[]) {
  ios::sync_with_stdio(false); cin.tie(NULL); // fast cin

  // read command line arguments - one parameter, WIDTH
  ll WIDTH = (argc > 1) ? (ll) stoi(argv[1]) : 79;

  // read input into paragraph blocks, maintaing empty lines
  ls lines;
  for (str line; getline(cin, line);) {
    if (len(line) > 0) {
      lines.append(line);
    }
    else {
      if (len(lines) > 0) {
        process(lines, WIDTH);
      }
      lines.clear();
      cout << el;
    }
  }
  if (len(lines) > 0) {
    process(lines, WIDTH);
  }

  return 0;
}

